04.ls/ls.rb

```diff
@@ -1,15 +1,96 @@
 #!/usr/bin/env ruby
 # frozen_string_literal: true
 
+require 'etc'
+
 require 'optparse'
 
 SPLIT_NUMBER = 3
 
+FILE_TYPE = {
+  '01' => 'p',
+  '02' => 'c',
+  '04' => 'd',
+  '06' => 'b',
+  '10' => '-',
+  '12' => 'l',
+  '14' => 's'
+}.freeze
+
+FILE_MODE = {
+  '0' => '---',
+  '1' => '--x',
+  '3' => '-w-',
+  '4' => 'r--',
+  '5' => 'r-x',
+  '6' => 'rw-',
+  '7' => 'rwx'
+}.freeze
+
 def main
-  options = ARGV.getopts('r')
+  options = ARGV.getopts('l')
   files = Dir.glob('*')
-  sorted_files = options['r'] ? files.reverse : files
-  display_files(sorted_files)
+  options['l'] ? display_long_format(files) : display_files(files)
+end
+
+def display_long_format(files)
+  long_formats = build_detailed_files(files)
+  total_block_number = calc_total_block_number(long_formats)
+  max_width = calc_max_width(long_formats)
+
+  puts "total #{total_block_number}"
+  long_formats.each do |format|
+    print "#{format[:file_mode]}  "
+    print "#{format[:hard_links_count].to_s.rjust(max_width[:link])} "
+    print "#{format[:owner].ljust(max_width[:owner])}  "
+    print "#{format[:group].ljust(max_width[:group])}  "
+    print "#{format[:file_size].to_s.rjust(max_width[:file_size])} "
+    print "#{format[:day_time]} "
+    print "#{format[:file_name]}\n"
+  end
+end
+
+def build_detailed_files(files)
+  files.map do |file|
+    file_info = File.stat(file)
+    {
+      block: file_info.blocks,
+      file_mode: fetch_file_mode(file_info),
+      hard_links_count: file_info.nlink,
+      owner: Etc.getpwuid(file_info.uid).name,
+      group: Etc.getgrgid(file_info.gid).name,
+      file_size: file_info.size,
+      day_time: file_info.mtime.strftime('%_m %_d %H:%M'),
+      file_name: File.basename(file)
+    }
+  end
+end
+
+def calc_total_block_number(long_formats)
+  block_long_formats = long_formats.map { |long_format| long_format[:block] }
+  block_long_formats.sum
+end
+
+def calc_max_width(long_formats)
+  {
+    link: long_formats.map { |long_format| long_format[:hard_links_count].to_s.size }.max,
+    owner: long_formats.map { |long_format| long_format[:owner].size }.max,
+    group: long_formats.map { |long_format| long_format[:group].size }.max,
+    file_size: long_formats.map { |long_format| long_format[:file_size].to_s.size }.max
+  }
+end
+
+def fetch_file_mode(file_info)
+  numerical_file_mode = file_info.mode.to_s(8)
+  symbolic_file_mode = convert_file_mode(numerical_file_mode.slice(3, 3)).join
+  symbolic_file_type = FILE_TYPE[numerical_file_mode.slice(0, 2)]
+  "#{symbolic_file_type}#{symbolic_file_mode}"
+end
+
+def convert_file_mode(file_mode_numbers)
+  file_mode_numbers.chars.map do |number|
+    FILE_MODE[number]
+  end
```

Comment:
高等テクニックですが正規表現が使えるかも？

```suggestion
  file_mode_numbers.gsub(/./, FILE_MODE)
```